os/drivers/input/ist415.c

```diff
@@ -89,185 +94,1899 @@ struct touchscreen_ops_s g_ist415_ops = {
+	} else {
+		memcpy(fw->buf, firmware.data, firmware.size);
+		fw->buf_size = firmware.size;
+	}
+
+	release_firmware(&firmware);
+
+	touchvdbg("firmware loaded successfully.\n");
+
+	ret = ist415_get_update_info(dev, fw->buf, fw->buf_size);
+	if (ret < 0) {
+		return ret;
+	}
+
+	fw->bin.main_ver = ist415_parse_ver(dev, FLAG_MAIN, fw->buf);
+	fw->bin.release_ver = ist415_parse_ver(dev, FLAG_RELEASE, fw->buf);
+	fw->bin.test_ver = ist415_parse_ver(dev, FLAG_TEST, fw->buf);
+	fw->bin.core_ver = ist415_parse_ver(dev, FLAG_CORE, fw->buf);
+	fw->bin.config_ver = ist415_parse_ver(dev, FLAG_CONFIG, fw->buf);
+
+	touchvdbg("Main : %08X Core: %x, Config %x, Release: %x, Test: %x\n", fw->bin.main_ver, fw->bin.core_ver, fw->bin.config_ver, fw->bin.release_ver, fw->bin.test_ver);
+
+	ret = ist415_check_auto_update(dev);
+	if (ret >= 0) {
+		return OK;
+	}
+
+	touchvdbg("Update version. release(core, config, test): %x(%x, %x, %x) -> %x(%x, %x, %x)\n", fw->cur.release_ver, fw->cur.core_ver, fw->cur.config_ver, fw->cur.test_ver, fw->bin.release_ver, fw->bin.core_ver, fw->bin.config_ver, fw->bin.test_ver);
+
+	retry = IST415_RETRY_CNT;
+	while (retry--) {
+		ret = ist415_fw_update(dev, fw->buf, fw->buf_size);
+		if (ret == OK) {
+			ist415_calibrate(dev, IST415_RETRY_CNT);
+			break;
+		}
+	}
+
+	return ret;
+}
+
+/****************************************************************************
+ * Name: ist415_print_info
+ ****************************************************************************/
+
+static void ist415_print_info(struct ist415_dev_s *dev)
+{
+	touchvdbg("*** TSP info ***\n");
+	touchvdbg(" swap: %d\n", dev->swap_xy);
+	touchvdbg(" tx: %d, rx: %d\n", dev->tx_num, dev->rx_num);
+	touchvdbg(" touch width: %d, height: %d\n", dev->width, dev->height);
+	touchvdbg(" mutual node len: %d\n", dev->mtl_node_len);
+	touchvdbg(" self node len: %d\n", dev->slf_node_len);
+	touchvdbg("Main Ver: %X, Test Ver: %X\n", dev->fw.cur.main_ver, dev->fw.cur.test_ver);
+	touchvdbg("Core Ver: %X, Config Ver: %X, Release Ver: %X\n", dev->fw.cur.core_ver, dev->fw.cur.config_ver, dev->fw.cur.release_ver);
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: ist415_close
+ ****************************************************************************/
+
+void ist415_close(struct ist415_dev_s *dev)
+{
+	FAR struct touchscreen_s *upper = dev->upper;
+
+	if (dev->knockknock) {
+		dev->sysmode = SYS_MODE_LPM;
+		ist415_power_mode(dev, dev->sysmode);
+	} else {
+		ist415_disable(upper);
+		touch_power_off(upper);
+	}
+}
+
+/****************************************************************************
+ * Name: ist415_open
+ ****************************************************************************/
+
+void ist415_open(struct ist415_dev_s *dev)
```

Comment:
these open/close never be used, actually touch_open toch_close using touch_enable/touch_disable.
Hence I think these logic need to be moved to ist415_enable / ist415_disable